### PR TITLE
fix: [CI-9262]: TI and log service to use url from delegate

### PR DESCRIPTION
**Pull request by smjt-h**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53050)

### CODE REVIEW

**Comment by smjt-h** [comment]
trigger uts

File: product/log-service/client/http.go [hunk 1]
```diff
@@ -241,6 +244,19 @@ func (c *HTTPClient) Info(ctx context.Context) (*stream.Info, error) {
...
+// Healthz returns the health information.
+func (c *HTTPClient) Healthz(ctx context.Context) error {
+	response, err := c.do(ctx, c.Endpoint+healthzEndpoint, "GET", nil, nil)
+	if err != nil {
+		return err
+	}
+	if response.StatusCode != http.StatusOK {
+		return fmt.Errorf("log-service Healthz Ping failed. Status Code:%s", response.Status)
+	}
+	fmt.Printf("log-service Healthz ping success")
```
Comment (rutvijmehta-harness) [inline product/log-service/client/http.go:256]: Do we need this log line? I don't think we print anything in the client.

File: product/ci/common/external/util.go [hunk 2]
```diff
@@ -173,18 +173,7 @@ func GetHTTPRemoteLogger(key string) (*logs.RemoteLogger, error) {
...
-		fmt.Println("%s env is set with value: %s ", delegateLogURLEnabledEnv, url)
-	} else {
-		url, ok = os.LookupEnv(logSvcEp)
-		if !ok {
-			return nil, fmt.Errorf("log service endpoint variable not set %s", logSvcEp)
-		}
-		fmt.Println("%s env is set with value from: %s : %s", logSvcEp, url)
-	}
-
+	certsDir := GetAdditionalCertsDir()
```
Comment (rutvijmehta-harness) [inline product/ci/common/external/util.go:176]: nit: This can be moved after L184

File: product/ci/common/external/util.go [hunk 3]
```diff
@@ -56,6 +55,7 @@ const (
 	harnessStageIndex        = "HARNESS_STAGE_INDEX"
 	harnessStageTotal        = "HARNESS_STAGE_TOTAL"
 	delegateLogURLEnabledEnv = "HARNESS_LE_DELEGATE_LOG_URL"
+	delegateTiURLEnabledEnv  = "HARNESS_LE_DELEGATE_TI_URL"
```
Comment (rutvijmehta-harness) [inline product/ci/common/external/util.go:58]: nit: `delegateLogUrlEnv` and `delegateTiUrlEnabledEnv`

**Comment by smjt-h** [comment]
trigger uts

File: product/ci/common/external/util.go (hunk 2 above)
Comment (smjt-h) [reply product/ci/common/external/util.go]: done

File: product/ci/common/external/util.go (hunk 3 above)
Comment (smjt-h) [reply product/ci/common/external/util.go]: done

File: product/log-service/client/http.go (hunk 1 above)
Comment (smjt-h) [reply product/log-service/client/http.go]: removed

File: product/ci/common/external/util.go [hunk 4]
```diff
@@ -232,7 +241,18 @@ func GetTiHTTPClient(repo, sha, commitLink string, skipVerify bool) ticlient.Cli
 	buildID, _ := GetBuildId()
 	stageID, _ := GetStageId()
 	certsDir := GetAdditionalCertsDir()
-	return ticlient.NewHTTPClient(endpoint, token, accountID, orgID, projectID, pipelineID, buildID, stageID, repo, sha,
+	delegateTiURL, delegateTiURLOk := os.LookupEnv(delegateTiURLEnabledEnv)
+	if delegateTiURLOk && strings.TrimSuffix(delegateTiURL, "/") != strings.TrimSuffix(tiEndpoint, "/") {
```
[the remark this answers is not in the document]
Comment (smjt-h) [reply product/ci/common/external/util.go]: done

**Comment by smjt-h** [comment]
trigger ut5

**Comment by smjt-h** [comment]
trigger ut3

**Comment by smjt-h** [comment]
trigger ut7

**Comment by smjt-h** [comment]
trigger uts

**Comment by smjt-h** [comment]
trigger uts

File: product/ci/common/external/util.go [hunk 5]
```diff
@@ -25,38 +24,39 @@ import (
...
+	diffFilesCmd       = "%s diff --name-status --diff-filter=MADR HEAD@{1} HEAD -1"
+	diffFilesCmdPush   = "%s diff --name-status --diff-filter=MADR %s"
+	safeDirCommand     = "%s config --global --add safe.directory '*'"
+	harnessStepIndex   = "HARNESS_STEP_INDEX"
+	harnessStepTotal   = "HARNESS_STEP_TOTAL"
+	harnessStageIndex  = "HARNESS_STAGE_INDEX"
+	harnessStageTotal  = "HARNESS_STAGE_TOTAL"
+	delegateLogURLEnv  = "HARNESS_LE_DELEGATE_LOG_URL"
+	delegateTiURLEnv   = "HARNESS_LE_DELEGATE_TI_URL"
+	LogHealthEndpoint  = "/healthz"
```
Comment (rutvijmehta-harness) [inline product/ci/common/external/util.go:59]: You can remove this now.

File: product/ci/common/external/util.go [hunk 6]
```diff
@@ -193,7 +180,28 @@ func GetRemoteHTTPClient() (client.Client, error) {
...
+		err := httpClient.Healthz(context.Background())
+		if err == nil {
+			fmt.Printf("%s env is set with value: %s \n", delegateLogURLEnv, delegateLogURL)
+			return httpClient, nil
+		} else {
+			fmt.Printf("Failed to ping log-service healthz: %w\n", err)
+		}
+	}
+	if internalLogURLOk {
+		fmt.Printf("Using internalLogURL:\n", internalLogURL)
```
Comment (rutvijmehta-harness) [inline product/ci/common/external/util.go:199]:
```suggestion
		fmt.Printf("Using internalLogURL: %s\n", internalLogURL)
```

File: product/log-service/client/http.go [hunk 7]
```diff
@@ -241,6 +244,18 @@ func (c *HTTPClient) Info(ctx context.Context) (*stream.Info, error) {
...
 }
 
+// Healthz returns the health information.
+func (c *HTTPClient) Healthz(ctx context.Context) error {
+	response, err := c.do(ctx, c.Endpoint+healthzEndpoint, "GET", nil, nil)
+	if err != nil {
+		return err
+	}
+	if response.StatusCode != http.StatusOK {
+		return fmt.Errorf("log-service Healthz Ping failed. Status Code:%s", response.Status)
```
Comment (rutvijmehta-harness) [inline product/log-service/client/http.go:254]:
```suggestion
		return fmt.Errorf("log-service Healthz Ping failed. Status Code:%s", response.StatusCode)
```

File: product/ci/common/external/util.go [hunk 8]
```diff
@@ -232,7 +240,18 @@ func GetTiHTTPClient(repo, sha, commitLink string, skipVerify bool) ticlient.Cli
...
 	stageID, _ := GetStageId()
 	certsDir := GetAdditionalCertsDir()
-	return ticlient.NewHTTPClient(endpoint, token, accountID, orgID, projectID, pipelineID, buildID, stageID, repo, sha,
+	delegateTiURL, delegateTiURLOk := os.LookupEnv(delegateTiURLEnv)
+	if delegateTiURLOk && !isUrlSame(delegateTiURL, tiEndpoint) {
+		tiHttpClient := ticlient.NewHTTPClient(delegateTiURL, token, accountID, orgID, projectID, pipelineID, buildID, stageID, repo, sha,
+			commitLink, skipVerify, certsDir)
+		err := tiHttpClient.Healthz(context.Background())
+		if err == nil {
+			fmt.Printf("%s env is set with value: %s \n", delegateTiURLEnv, delegateTiURL)
```
Comment (rutvijmehta-harness) [inline product/ci/common/external/util.go:249]:
```suggestion
			fmt.Printf("%s env is set with value: %s\n", delegateTiURLEnv, delegateTiURL)
```

File: product/ci/common/external/util.go [hunk 9]
```diff
@@ -232,7 +240,18 @@ func GetTiHTTPClient(repo, sha, commitLink string, skipVerify bool) ticlient.Cli
...
+			return tiHttpClient
+		}
```
Comment (rutvijmehta-harness) [inline product/ci/common/external/util.go:251]: If `err != nil`, add a similar log line `fmt.Printf("Failed to ping ti-service: %w\n", err)`

File: product/ci/common/external/util.go [hunk 10]
```diff
@@ -193,7 +180,28 @@ func GetRemoteHTTPClient() (client.Client, error) {
...
+	internalLogURL, internalLogURLOk := os.LookupEnv(logSvcEp)
+
+	if delegateLogURLOk && !isUrlSame(delegateLogURL, internalLogURL) {
+		httpClient := client.NewHTTPClient(delegateLogURL, account, token, false, certsDir)
+		err := httpClient.Healthz(context.Background())
+		if err == nil {
+			fmt.Printf("%s env is set with value: %s \n", delegateLogURLEnv, delegateLogURL)
+			return httpClient, nil
+		} else {
+			fmt.Printf("Failed to ping log-service healthz: %w\n", err)
```
Comment (rutvijmehta-harness) [inline product/ci/common/external/util.go:195]:
```suggestion
			fmt.Printf("Failed to ping log-service: %w\n", err)
```

File: product/ci/common/external/util.go (hunk 5 above)
Comment (smjt-h) [reply product/ci/common/external/util.go:59]: done

**Comment by smjt-h** [comment]
trigger uts

**Review by rutvijmehta-harness — APPROVED** [review]

**Pull request merged**